src/formats/CIF.cpp

```diff
@@ -0,0 +1,181 @@
+#include "chemfiles/utils.hpp"
+#include "chemfiles/parse.hpp"
+#include "chemfiles/error_fmt.hpp"
+#include "chemfiles/string_view.hpp"
+#include "chemfiles/external/optional.hpp"
+
+#include "chemfiles/formats/CIF.hpp"
+
+using namespace chemfiles;
+
+template<> FormatInfo chemfiles::format_information<CIFFormat>() {
+    return FormatInfo("CIF").with_extension(".cif").description(
+        "CIF (Crystallographic Information Framework)"
+    );
+}
+
+void CIFFormat::init_() {
+    gemmi::cif::Document *doc;
+
+    if (file_.mode() == File::WRITE) {
+        return;
+    }
+
+    if (file_.mode() == File::APPEND) {
+        throw file_error("cannot open CIF files in append ('a') mode");
+    }
+
+    // Parse the CIF file
+    auto content = file_.readall();
+    try {
+	doc = new gemmi::cif::Document(gemmi::cif::read_string(content));
+    } catch (std::exception& e) {
+	throw format_error("cannot parse CIF file");
+    }
+
+    // Interpret the content of each block, to remove blocks that contain
+    // no atom: those can occur, but are not useful to us (they contain comments
+    // and associated experimental data).
+    for (const auto &block: doc->blocks) {
+      try {
+	gemmi::SmallStructure s = gemmi::make_small_structure_from_block(block);
+
+	if (s.get_all_unit_cell_sites().size() > 0) {
+	  structures_.push_back(s);
+	}
+      } catch (std::exception& e) {
+	/* Simply skip to the next block. */
+	throw format_error("cannot interpret CIF block");
+      }
+    }
+
+    // TODO: if the CIF file has zero blocks with atoms, maybe we should throw an error here
+    // if (structures_.size() == 0) ...
+}
+
+size_t CIFFormat::nsteps() {
+    return structures_.size();
+}
+
+void CIFFormat::read_step(const size_t step, Frame& frame) {
+    assert(step < structures_.size());
+
+    auto structure = structures_[step];
+    auto sites = structure.get_all_unit_cell_sites();
+
+    UnitCell cell = UnitCell(structure.cell.a, structure.cell.b, structure.cell.c,
+	structure.cell.alpha, structure.cell.beta, structure.cell.gamma);
+    frame.set_cell(cell);
+
+    frame.set("name", structure.name);
+
+    for (const auto& site: sites) {
+        Atom atom(site.label, site.element.name());
+        // TODO: should we use those somehow?
+	// site.element.atomic_number()
+	// site.element.weight()
+	// site.occ is the occupancy
+
+	gemmi::Fractional fract = site.fract.wrap_to_unit();
+	gemmi::Position p = structure.cell.orthogonalize(fract);
+
+        frame.add_atom(std::move(atom), Vector3D(p.x, p.y, p.z));
+    }
+}
+
+// TODO: is it mandatory to implement both read_step() and read()?
+void CIFFormat::read(Frame& frame) {
+    read_step(current_step_, frame);
+    current_step_++;
+}
+
+void CIFFormat::write(const Frame& frame) {
+    auto name = frame.get("name");
+    if (name && name->kind() == Property::STRING) {
+        file_.print("data_{}\n", name->as_string());
+    } else {
+        file_.print("data_model_{}\n", models_);
```

Comment:
seems like this was taken from mmCIF, which uses `model` as the trajectory step. Instead we could use `current_step_` here